Makefile

```diff
@@ -193,9 +193,11 @@ cmd/promtail/promtail-debug: $(APP_GO_FILES) pkg/promtail/server/ui/assets_vfsda
 #############
 # concurrency is limited to 4 to prevent CircleCI from OOMing. Sorry
 GOX = gox $(GO_FLAGS) -parallel=4 -output="dist/{{.Dir}}-{{.OS}}-{{.Arch}}" -arch="amd64 arm64 arm" -os="linux"
+CGO_GOX = gox $(DYN_GO_FLAGS) -output="dist/{{.Dir}}-{{.OS}}-{{.Arch}}" -arch="amd64" -os="linux"
 dist: clean
 	CGO_ENABLED=0 $(GOX) -osarch="windows/amd64" ./cmd/loki
 	CGO_ENABLED=0 $(GOX) -osarch="darwin/amd64 windows/amd64 freebsd/amd64" ./cmd/promtail ./cmd/logcli
+	CGO_ENABLED=1 $(CGO_GOX) ./cmd/promtail
```

Comment:
I think to me this might be easier to read if we break things up:

```
GOX = gox $(GO_FLAGS) -parallel=4 -output="dist/{{.Dir}}-{{.OS}}-{{.Arch}}" 
CGO_GOX = gox $(DYN_GO_FLAGS) -output="dist/{{.Dir}}-{{.OS}}-{{.Arch}}" 
dist: clean
	CGO_ENABLED=0 $(GOX) -osarch="linux/amd64 linux/arm64 linux/arm windows/amd64" ./cmd/loki
	CGO_ENABLED=0 $(GOX) -osarch="linux/amd64 linux/arm64 linux/arm windows/amd64 darwin/amd64 freebsd/amd64" ./cmd/logcli
	CGO_ENABLED=0 $(GOX) -osarch="linux/arm64 linux/arm windows/amd64 darwin/amd64 freebsd/amd64" ./cmd/promtail
	CGO_ENABLED=1 $(CGO_GOX) -osarch="linux/amd64" ./cmd/promtail
```

Reply:
Good idea, done.

Reply:
I know that it was a long time ago, but I have a few silly questions here ))) 

## Context

promtail binary does not work on OpenWRT (https://github.com/grafana/loki/issues/7764)

## Questions
1. Can we provide a few variants of promtail binary: "dynamic" and "static" but both with CGO_ENABLED=1?
2. I see that we build promtail `CGO_ENABLED=0 $(GOX) -osarch="linux/arm64 linux/arm windows/amd64 darwin/amd64 freebsd/amd64" ./cmd/promtail` without CGO and after this we run `CGO_ENABLED=1 $(CGO_GOX) -osarch="linux/amd64" ./cmd/promtail` that builds promtail using GCO and overrides promtail binary for `linux/amd64` (because the names of the binaries are the same). 
Why do we need to use such an override? can we skip `linux/arm64` variant when we build promtail with disabled CGO but build a variant for this os/arch with enabled CGO?
something like this:
`CGO_ENABLED=0 $(GOX) -osarch="linux/arm windows/amd64 darwin/amd64 freebsd/amd64" ./cmd/promtail`
`CGO_ENABLED=1 $(CGO_GOX) -osarch="linux/amd64" ./cmd/promtail`
3. why do we actually need dynamic binaries? I see that it was added in [this PR](https://github.com/grafana/loki/pull/809) but I do not know why we can not just build a static binary that will include all required pages....

Reply:
I'll try to remember, it has been a very long time :) 

> Can we provide a few variants of promtail binary: "dynamic" and "static" but both with CGO_ENABLED=1?

Probably. At the time, my opinion was probably that it would be confusing to have more than one Promtail binary with different features supported. 

> I see that we build promtail [...] without CGO and after this we [build] promtail using GCO and overrides promtail binary for linux/amd64 (because the names of the binaries are the same

Are you sure it's overriding? The first command without CGO doesn't build for linux/amd64 at all, just linux/arm64. 

> why do we actually need dynamic binaries? I see that it was added in https://github.com/grafana/loki/pull/809 but I do not know why we can not just build a static binary that will include all required libs....

From what I can remember, it's a glibc issue, but I can't remember the details. I got some kind of warning or error message when I tried building a static binary. 

Plus, you can't build a fully static library; systemd is only available as a dynamic library and can't be linked statically.

Reply:
> Are you sure it's overriding? The first command without CGO doesn't build for linux/amd64 at all, just linux/arm64.

You are right ;) my bad.

> From what I can remember, it's a glibc issue, but I can't remember the details. I got some kind of warning or error message when I tried building a static binary.
Plus, you can't build a fully static library; systemd is only available as a dynamic library and can't be linked statically.

I see. thanks a lot )